Allow `export_report` to run without waveforms

chrishalcrow opened the pull request:
Small edit to `export_report` so that it will generate a report without waveforms.

When we're being speedy, we often don't compute or don't copy over waveforms. This PR just changes `export_report` so that it does/doesn't include the `UnitWaveformDensityMapWidget` if the sorting_analyzer does/doesn't have waveforms computed.

A unit report without waveforms:

![4](https://github.com/user-attachments/assets/cbbf296a-b573-47b8-9c49-cd8b91978b0d)

computed using

```
import spikeinterface.full as si
rec, sort = si.generate_ground_truth_recording()
sa = si.create_sorting_analyzer(recording=rec, sorting=sort)
sa.compute(["random_spikes",  "templates", "spike_amplitudes", "correlograms", "template_similarity"])
si.export_report(sorting_analyzer=sa, output_folder="my_report")
```

zm711 commented:
I think this is an improvement. Another request we should think about is to have more flexibility in selecting what to plot in general. There are complaints about the slowness of making the summary when you only want some features to be saved. So something like this could be extended to help speed things up for people.

samuelgarcia commented:
Good idea. thanks Chris.

chrishalcrow commented:
> I think this is an improvement. Another request we should think about is to have more flexibility in selecting what to plot in general. There are complaints about the slowness of making the summary when you only want some features to be saved. So something like this could be extended to help speed things up for people.

Mmm, good idea. So at the moment we include 5 widgets. Your suggestion would be to allow the user to select which ones they'd like, and we format the report appropriately? Sounds doable (and extendable for other widgets!) in another PR.

zm711 commented:
> in another PR.

agreed. :)

JoeZiminski commented on src/spikeinterface/widgets/unit_summary.py:
```diff
@@ -136,37 +140,41 @@ def plot_matplotlib(self, data_plot, **backend_kwargs):
+                **unitwaveformdensitymapwidget_kwargs,
+            )
+            ax3.set_ylabel(None)
+            col_counter = col_counter + 1
```
these could be `col_counter += 1`

chrishalcrow replied:
I know that `c++` isn't in python, and this somehow merged in my mind with the `+=` operator. Omg I've been coding in python for 9 months now... thanks Joe!

JoeZiminski commented on src/spikeinterface/widgets/unit_summary.py:
```diff
@@ -107,15 +107,18 @@ def plot_matplotlib(self, data_plot, **backend_kwargs):
 
         if sorting_analyzer.has_extension("unit_locations"):
-            ax1 = fig.add_subplot(gs[:2, 0])
+            ax1 = fig.add_subplot(gs[:2, col_counter])
```
ATM this is `ax1` and the next axis is `ax2` but if `unit_locations` is not found then `ax2` will actually be the first axis. Is it worth naming these by their plot type? e.g. `unit_locations_axis` (but maybe this is confusing in other ways as you loose any sense of order)

JoeZiminski commented on src/spikeinterface/widgets/unit_summary.py:
```diff
@@ -107,15 +107,18 @@ def plot_matplotlib(self, data_plot, **backend_kwargs):
         fig = self.figure
         nrows = 2
-        ncols = 3
+        ncols = 2
```
It might be worth expanding the docs just to briefly introduce what things are plot and under what conditions. For example are some plots conditional on the steps run while others will always be plot? I got a little confused here because `unit_locations` is under a conditional but it's inclusion doesn't add to the number `ncols`. Maybe the initial `ncols=2` is reserved for `unit_locations`? but then I don't think `unit_locations` need to be under the conditional if it will always be plot. I tried skipping unit locations and there was some unused space in the plot so maybe this is the case 🤔

JoeZiminski replied:
(I just read the diff more carefully and see this is not actually related to your changes in this PR, sorry! but maybe you know the answer

chrishalcrow replied:
Hello - just noticed that at the moment the `export_report` function computes `unit_locations` if they're not already computed (line 90 in `report.py`). So the first plot also always appears.

So either:
- remove the conditional on the unit locations plot
- remove the automatic computation

Opinions @samuelgarcia @zm711 ?
I'd vote to remove the automatic computation, since I don't think we should be computing anything in a plotting function.

chrishalcrow replied:
Sounds ok to me, Zach! I've updated to remove the conditional on the unit locations plot.

JoeZiminski commented on src/spikeinterface/widgets/unit_summary.py:
```diff
@@ -136,37 +140,41 @@ def plot_matplotlib(self, data_plot, **backend_kwargs):
             unit_ids=[unit_id],
             unit_colors=unit_colors,
             plot_templates=True,
+            plot_waveforms=sorting_analyzer.has_extension("waveforms"),
```
If `UnitWaveformsWidget` is the only plot that is always plot, could it be moved to the top of this section? This would help make it clearer what plots are conditional and what plots are always plot (i.e. always plot at the top, then conditional).

JoeZiminski replied:
Actually I guess this doesn't work if this plot should be after the first `UnitWaveformsWidget` (if it exists). Maybe a comment above like `#This plot is always displayed, if 'unit_locations' exists it will be plot in the column before it`.

JoeZiminski commented on src/spikeinterface/widgets/unit_summary.py:
```diff
@@ -107,15 +107,18 @@ def plot_matplotlib(self, data_plot, **backend_kwargs):
+        ncols = 2
         if sorting_analyzer.has_extension("correlograms") or sorting_analyzer.has_extension("spike_amplitudes"):
             ncols += 1
+        if sorting_analyzer.has_extension("waveforms"):
```
Sorry not related to this line but just above, two small things not from this diff but now seems a good time to fix:

- `            # UnitLocationsPlotter().do_plot(dp.plot_data_unit_locations, ax=ax1)` this commented line could be removed
- the variables `w` as output from both calls to `UnitLocationsWidget` is unused and could be removed.

JoeZiminski reviewed (approved):
Nice @chrishalcrow this is really cool, the export report is a great feature and making it more flexible is very useful. Also agree that custom selecting of subplots would be very nice (in another PR). 

I ran it and it all worked nicely. Some small changes suggested, apologies some are outside this diff as I was reading the code in another editor, but I guess now would be a good time to change them.

chrishalcrow commented:
Thanks @JoeZiminski - will take a look on Monday!

chrishalcrow commented:
All updated - also fixed a column counting issue that occurred when correlograms and spike_amplitudes were on, but waveforms were off. Also added in a check for `template_similarity`, which is needed to compute correlograms it seems.

zm711 commented:
> template_similarity, which is needed to compute correlograms it seems.

yeah the correlograms are only calculated for templates deemed similar enough. That's annoyed me in the past, but I understand why it is there.

chrishalcrow commented:
Hello, could this be merged please(@samuelgarcia )? Just to sum up, here is every possible report output, made from

```
import spikeinterface.full as si
rec, sort = si.generate_ground_truth_recording(durations=[60])
sa = si.create_sorting_analyzer(recording=rec, sorting=sort)
```
then...

`sa.compute(["random_spikes", "unit_locations","templates","template_similarity"])`
![0](https://github.com/user-attachments/assets/257a26ca-8ff9-41cf-9527-4b9155df4ff3)

`sa.compute(["random_spikes", "unit_locations", "templates","template_similarity", "correlograms"])`
![0](https://github.com/user-attachments/assets/f528aa4f-59dc-4d76-b6db-daefb47fdaf0)

`sa.compute(["random_spikes", "unit_locations", "templates","template_similarity", "waveforms"])`
![0](https://github.com/user-attachments/assets/dfb7d228-9acd-43e8-bb4c-dc5dc61e952d)

`sa.compute(["random_spikes", "unit_locations", "templates","template_similarity", "spike_amplitudes"])`
![0](https://github.com/user-attachments/assets/f14019af-5aa4-488a-955d-c13cbd8487b9)

`sa.compute(["random_spikes", "unit_locations", "templates", "spike_amplitudes", "correlograms"])`
(note: no template similarity => no correlograms!)
![0](https://github.com/user-attachments/assets/14f9eddb-0973-46ae-917d-96136a72ed86)

`sa.compute(["random_spikes", "unit_locations","templates","template_similarity", "waveforms", "correlograms"])`
![0](https://github.com/user-attachments/assets/30c0d747-ded3-4d2a-bab7-16e49b7296c8)

`sa.compute(["random_spikes", "unit_locations","templates","template_similarity", "waveforms", "spike_amplitudes"])`
![0](https://github.com/user-attachments/assets/4575428e-c9dc-454d-8c6a-54477dc71758)

`sa.compute(["random_spikes", "unit_locations","templates","template_similarity", "waveforms", "correlograms", "spike_amplitudes"])`
![0](https://github.com/user-attachments/assets/bd6e5e59-bd93-4656-8c98-ac7e5194e0b0)

alejoe91 commented:
@chrishalcrow @zm711 I pushed an update that removes the need to compute `template_similarity`, which is counter-intuituve. In fact, this is only needed for cross-correlograms, but not for auto-correlogram plots.

Tested on my side and works like a charm!

chrishalcrow commented:
> @chrishalcrow @zm711 I pushed an update that removes the need to compute `template_similarity`, which is counter-intuituve. In fact, this is only needed for cross-correlograms, but not for auto-correlogram plots.
> 
> Tested on my side and works like a charm!

Beautiful - works on my side too.
